Cannot set sensor orientation to YAW_45

### Issue details

I'm trying to use a rangefinder (distance sensor) at a 45° angle in SITL with Gazebo. I've successfully set orientation to NONE, YAW_90, YAW_180, YAW_270.
For example for YAW_90, the sensor definition in my .sdf file is :
```yaml
<!--right-facing lidar-->
    <include>
      <uri>model://lidar</uri>
      <pose>0 0 -0.1 0 0 -1.57079633</pose>
      <name>lidar1</name>
    </include>
``` 

Similarly for YAW_45, I used :
```yaml
<!--45°-facing lidar-->
    <include>
      <uri>model://lidar</uri>
      <pose>0 0 -0.1 0 0 0.785398163</pose>
      <name>lidar4</name>
    </include>
``` 

But I get the following error : `SENSOR: wrong orientation index: 100`
From what I understood of https://github.com/mavlink/mavros/blob/master/mavros/src/lib/enum_sensor_orientation.cpp my orientation is considered as CUSTOM while it should be considered as YAW_45. Maybe this is a deg/rad conversion error or approximation. I've tried with several approximations of π/4.



### MAVROS version and platform

Mavros: latest
ROS: Noetic
Ubuntu: 20.04


### Autopilot type and version

[  ] ArduPilot
[ X ] PX4

Version: latest


### Node logs

```
[ERROR] [1641489853.571286350, 184.549000000]: SENSOR: wrong orientation index: 100
```

### Diagnostics

```
header: 
  seq: 15
  stamp: 
    secs: 193
    nsecs: 938000000
  frame_id: ''
status: 
  - 
    level: 0
    name: "mavros: FCU connection"
    message: "connected"
    hardware_id: "udp://:14540@localhost:14557"
    values: 
      - 
        key: "Received packets:"
        value: "3540"
      - 
        key: "Dropped packets:"
        value: "0"
      - 
        key: "Buffer overruns:"
        value: "0"
      - 
        key: "Parse errors:"
        value: "0"
      - 
        key: "Rx sequence number:"
        value: "230"
      - 
        key: "Tx sequence number:"
        value: "0"
      - 
        key: "Rx total bytes:"
        value: "206705"
      - 
        key: "Tx total bytes:"
        value: "5667"
      - 
        key: "Rx speed:"
        value: "27933.000000"
      - 
        key: "Tx speed:"
        value: "537.000000"
  - 
    level: 0
    name: "mavros: GPS"
    message: "3D fix"
    hardware_id: "udp://:14540@localhost:14557"
    values: 
      - 
        key: "Satellites visible"
        value: "10"
      - 
        key: "Fix type"
        value: "3"
      - 
        key: "EPH (m)"
        value: "0.00"
      - 
        key: "EPV (m)"
        value: "0.00"
  - 
    level: 1
    name: "mavros: Heartbeat"
    message: "Frequency too low."
    hardware_id: "udp://:14540@localhost:14557"
    values: 
      - 
        key: "Heartbeats since startup"
        value: "9"
      - 
        key: "Frequency (Hz)"
        value: "0.046407"
      - 
        key: "Vehicle type"
        value: "Quadrotor"
      - 
        key: "Autopilot type"
        value: "PX4 Autopilot"
      - 
        key: "Mode"
        value: "POSCTL"
      - 
        key: "System status"
        value: "Standby"
  - 
    level: 0
    name: "mavros: System"
    message: "Normal"
    hardware_id: "udp://:14540@localhost:14557"
    values: 
      - 
        key: "Sensor present"
        value: "0x3029002C"
      - 
        key: "Sensor enabled"
        value: "0x3021000C"
      - 
        key: "Sensor health"
        value: "0x302F002F"
      - 
        key: "3D magnetometer"
        value: "Ok"
      - 
        key: "absolute pressure"
        value: "Ok"
      - 
        key: "rc receiver"
        value: "Ok"
      - 
        key: "AHRS subsystem health"
        value: "Ok"
      - 
        key: "pre-arm check status. Always healthy when armed"
        value: "Ok"
      - 
        key: "Avoidance/collision prevention"
        value: "Ok"
      - 
        key: "CPU Load (%)"
        value: "28.0"
      - 
        key: "Drop rate (%)"
        value: "0.0"
      - 
        key: "Errors comm"
        value: "0"
      - 
        key: "Errors count #1"
        value: "0"
      - 
        key: "Errors count #2"
        value: "0"
      - 
        key: "Errors count #3"
        value: "0"
      - 
        key: "Errors count #4"
        value: "0"
  - 
    level: 0
    name: "mavros: Battery"
    message: "Normal"
    hardware_id: "udp://:14540@localhost:14557"
    values: 
      - 
        key: "Voltage"
        value: "16.20"
      - 
        key: "Current"
        value: "-1.0"
      - 
        key: "Remaining"
        value: "100.0"
  - 
    level: 0
    name: "mavros: Time Sync"
    message: "Normal"
    hardware_id: "udp://:14540@localhost:14557"
    values: 
      - 
        key: "Timesyncs since startup"
        value: "96"
      - 
        key: "Frequency (Hz)"
        value: "0.495004"
      - 
        key: "Last RTT (ms)"
        value: "1.000000"
      - 
        key: "Mean RTT (ms)"
        value: "0.229166"
      - 
        key: "Last remote time (s)"
        value: "9.764000000"
      - 
        key: "Estimated time offset (s)"
        value: "184.144288028"
---
```

### Check ID

```
OK. I got messages from 1:1.

---
Received 7420 messages, from 2 addresses
sys:comp   list of messages
Exception in thread Thread-5:
Traceback (most recent call last):
  File "/usr/lib/python3.8/threading.py", line 932, in _bootstrap_inner
    self.run()
  File "/opt/ros/noetic/lib/python3/dist-packages/rospy/timer.py", line 237, in run
    self._callback(TimerEvent(last_expected, last_real, current_expected, current_real, last_duration))
  File "/home/mathieu/catkin_ws/src/mavros/mavros/scripts/checkid", line 82, in timer_cb
    for address, messages in self.message_sources.iteritems():
AttributeError: 'dict' object has no attribute 'iteritems'
```

My config for this sensor is :
```yaml
rangefinder_pub4:
    id: 4
    frame_id: "lidar4"
    send_tf: true
    sensor_position: {x:  0.0, y:  0.0, z:  -0.1}
    field_of_view: 0.0
```

I've also tried with orientation :
```yaml
rangefinder_pub4:
    id: 4
    orientation: YAW_45
    frame_id: "lidar4"
    send_tf: true
    sensor_position: {x:  0.0, y:  0.0, z:  -0.1}
    field_of_view: 0.0
```
but in this case I get another error : 
```
[ERROR] [1641647988.732717328, 4.408000000]: SENSOR: wrong orientation index: 100
[ERROR] [1641647988.732796576, 4.408000000]: DS: rangefinder_pub4: received sensor data has different orientation (100) than in config (YAW_45)!
```
This is is what gave me the impression it read the SDF file.

--- a/mavros/scripts/checkid
+++ b/mavros/scripts/checkid
@@ -79,7 +79,7 @@ class Checker(object):
 
         # TODO: prettytable?
         print("sys:comp   list of messages")
-        for address, messages in self.message_sources.iteritems():
+        for address, messages in self.message_sources.items():
             print("% 3d:%-3d   %s" % (
                 address[0], address[1],
                 ", ".join(("%d" % msgid for msgid in messages))

--- a/mavros/src/lib/enum_sensor_orientation.cpp
+++ b/mavros/src/lib/enum_sensor_orientation.cpp
@@ -14,13 +14,14 @@
  * https://github.com/mavlink/mavros/tree/master/LICENSE.md
  */
 
+#include <unordered_map>
+
 #include <mavros/utils.h>
 #include <mavros/frame_tf.h>
 #include <ros/console.h>
 
 namespace mavros {
 namespace utils {
-
 using mavlink::common::MAV_SENSOR_ORIENTATION;
 
 // internal type: name - rotation
@@ -64,83 +65,86 @@ static const OrientationPair make_orientation(const std::string &name,
 //             print(f"Parse Error: {ex}, desc: {desc}")
 //             return cls()
 //
-// cog.outl("static const std::array<const OrientationPair, %s> sensor_orientations{{" % len(enum))
+// cog.outl(f"static const std::unordered_map<typename std::underlying_type<{ename}>::type, const OrientationPair> sensor_orientations{{{{")
 // for k, e in enum:
 //     name_short = e.name[len(pfx2):]
 //     vec = Vector3.parse_rpy(e.description)
 //     whitespace = ' ' * (27 - len(name_short))
-//     cog.outl(f"""/* {k:>2} */ make_orientation("{name_short}",{whitespace}{vec.Roll:>5}, {vec.Pitch:>5}, {vec.Yaw:>5}),""")
+//     cog.outl(f"""{{ {k:>3}, make_orientation("{name_short}",{whitespace}{vec.Roll:>5}, {vec.Pitch:>5}, {vec.Yaw:>5}) }},""")
 //
 // cog.outl("}};")
 // ]]]
-static const std::array<const OrientationPair, 42> sensor_orientations{{
-/*  0 */ make_orientation("NONE",                         0.0,   0.0,   0.0),
-/*  1 */ make_orientation("YAW_45",                       0.0,   0.0,  45.0),
-/*  2 */ make_orientation("YAW_90",                       0.0,   0.0,  90.0),
-/*  3 */ make_orientation("YAW_135",                      0.0,   0.0, 135.0),
-/*  4 */ make_orientation("YAW_180",                      0.0,   0.0, 180.0),
-/*  5 */ make_orientation("YAW_225",                      0.0,   0.0, 225.0),
-/*  6 */ make_orientation("YAW_270",                      0.0,   0.0, 270.0),
-/*  7 */ make_orientation("YAW_315",                      0.0,   0.0, 315.0),
-/*  8 */ make_orientation("ROLL_180",                   180.0,   0.0,   0.0),
-/*  9 */ make_orientation("ROLL_180_YAW_45",            180.0,   0.0,  45.0),
-/* 10 */ make_orientation("ROLL_180_YAW_90",            180.0,   0.0,  90.0),
-/* 11 */ make_orientation("ROLL_180_YAW_135",           180.0,   0.0, 135.0),
-/* 12 */ make_orientation("PITCH_180",                    0.0, 180.0,   0.0),
-/* 13 */ make_orientation("ROLL_180_YAW_225",           180.0,   0.0, 225.0),
-/* 14 */ make_orientation("ROLL_180_YAW_270",           180.0,   0.0, 270.0),
-/* 15 */ make_orientation("ROLL_180_YAW_315",           180.0,   0.0, 315.0),
-/* 16 */ make_orientation("ROLL_90",                     90.0,   0.0,   0.0),
-/* 17 */ make_orientation("ROLL_90_YAW_45",              90.0,   0.0,  45.0),
-/* 18 */ make_orientation("ROLL_90_YAW_90",              90.0,   0.0,  90.0),
-/* 19 */ make_orientation("ROLL_90_YAW_135",             90.0,   0.0, 135.0),
-/* 20 */ make_orientation("ROLL_270",                   270.0,   0.0,   0.0),
-/* 21 */ make_orientation("ROLL_270_YAW_45",            270.0,   0.0,  45.0),
-/* 22 */ make_orientation("ROLL_270_YAW_90",            270.0,   0.0,  90.0),
-/* 23 */ make_orientation("ROLL_270_YAW_135",           270.0,   0.0, 135.0),
-/* 24 */ make_orientation("PITCH_90",                     0.0,  90.0,   0.0),
-/* 25 */ make_orientation("PITCH_270",                    0.0, 270.0,   0.0),
-/* 26 */ make_orientation("PITCH_180_YAW_90",             0.0, 180.0,  90.0),
-/* 27 */ make_orientation("PITCH_180_YAW_270",            0.0, 180.0, 270.0),
-/* 28 */ make_orientation("ROLL_90_PITCH_90",            90.0,  90.0,   0.0),
-/* 29 */ make_orientation("ROLL_180_PITCH_90",          180.0,  90.0,   0.0),
-/* 30 */ make_orientation("ROLL_270_PITCH_90",          270.0,  90.0,   0.0),
-/* 31 */ make_orientation("ROLL_90_PITCH_180",           90.0, 180.0,   0.0),
-/* 32 */ make_orientation("ROLL_270_PITCH_180",         270.0, 180.0,   0.0),
-/* 33 */ make_orientation("ROLL_90_PITCH_270",           90.0, 270.0,   0.0),
-/* 34 */ make_orientation("ROLL_180_PITCH_270",         180.0, 270.0,   0.0),
-/* 35 */ make_orientation("ROLL_270_PITCH_270",         270.0, 270.0,   0.0),
-/* 36 */ make_orientation("ROLL_90_PITCH_180_YAW_90",    90.0, 180.0,  90.0),
-/* 37 */ make_orientation("ROLL_90_YAW_270",             90.0,   0.0, 270.0),
-/* 38 */ make_orientation("ROLL_90_PITCH_68_YAW_293",    90.0,  68.0, 293.0),
-/* 39 */ make_orientation("PITCH_315",                    0.0, 315.0,   0.0),
-/* 40 */ make_orientation("ROLL_90_PITCH_315",           90.0, 315.0,   0.0),
-/* 100 */ make_orientation("CUSTOM",                       0.0,   0.0,   0.0),
+static const std::unordered_map<typename std::underlying_type<MAV_SENSOR_ORIENTATION>::type, const OrientationPair> sensor_orientations{{
+{   0, make_orientation("NONE",                         0.0,   0.0,   0.0) },
+{   1, make_orientation("YAW_45",                       0.0,   0.0,  45.0) },
+{   2, make_orientation("YAW_90",                       0.0,   0.0,  90.0) },
+{   3, make_orientation("YAW_135",                      0.0,   0.0, 135.0) },
+{   4, make_orientation("YAW_180",                      0.0,   0.0, 180.0) },
+{   5, make_orientation("YAW_225",                      0.0,   0.0, 225.0) },
+{   6, make_orientation("YAW_270",                      0.0,   0.0, 270.0) },
+{   7, make_orientation("YAW_315",                      0.0,   0.0, 315.0) },
+{   8, make_orientation("ROLL_180",                   180.0,   0.0,   0.0) },
+{   9, make_orientation("ROLL_180_YAW_45",            180.0,   0.0,  45.0) },
+{  10, make_orientation("ROLL_180_YAW_90",            180.0,   0.0,  90.0) },
+{  11, make_orientation("ROLL_180_YAW_135",           180.0,   0.0, 135.0) },
+{  12, make_orientation("PITCH_180",                    0.0, 180.0,   0.0) },
+{  13, make_orientation("ROLL_180_YAW_225",           180.0,   0.0, 225.0) },
+{  14, make_orientation("ROLL_180_YAW_270",           180.0,   0.0, 270.0) },
+{  15, make_orientation("ROLL_180_YAW_315",           180.0,   0.0, 315.0) },
+{  16, make_orientation("ROLL_90",                     90.0,   0.0,   0.0) },
+{  17, make_orientation("ROLL_90_YAW_45",              90.0,   0.0,  45.0) },
+{  18, make_orientation("ROLL_90_YAW_90",              90.0,   0.0,  90.0) },
+{  19, make_orientation("ROLL_90_YAW_135",             90.0,   0.0, 135.0) },
+{  20, make_orientation("ROLL_270",                   270.0,   0.0,   0.0) },
+{  21, make_orientation("ROLL_270_YAW_45",            270.0,   0.0,  45.0) },
+{  22, make_orientation("ROLL_270_YAW_90",            270.0,   0.0,  90.0) },
+{  23, make_orientation("ROLL_270_YAW_135",           270.0,   0.0, 135.0) },
+{  24, make_orientation("PITCH_90",                     0.0,  90.0,   0.0) },
+{  25, make_orientation("PITCH_270",                    0.0, 270.0,   0.0) },
+{  26, make_orientation("PITCH_180_YAW_90",             0.0, 180.0,  90.0) },
+{  27, make_orientation("PITCH_180_YAW_270",            0.0, 180.0, 270.0) },
+{  28, make_orientation("ROLL_90_PITCH_90",            90.0,  90.0,   0.0) },
+{  29, make_orientation("ROLL_180_PITCH_90",          180.0,  90.0,   0.0) },
+{  30, make_orientation("ROLL_270_PITCH_90",          270.0,  90.0,   0.0) },
+{  31, make_orientation("ROLL_90_PITCH_180",           90.0, 180.0,   0.0) },
+{  32, make_orientation("ROLL_270_PITCH_180",         270.0, 180.0,   0.0) },
+{  33, make_orientation("ROLL_90_PITCH_270",           90.0, 270.0,   0.0) },
+{  34, make_orientation("ROLL_180_PITCH_270",         180.0, 270.0,   0.0) },
+{  35, make_orientation("ROLL_270_PITCH_270",         270.0, 270.0,   0.0) },
+{  36, make_orientation("ROLL_90_PITCH_180_YAW_90",    90.0, 180.0,  90.0) },
+{  37, make_orientation("ROLL_90_YAW_270",             90.0,   0.0, 270.0) },
+{  38, make_orientation("ROLL_90_PITCH_68_YAW_293",    90.0,  68.0, 293.0) },
+{  39, make_orientation("PITCH_315",                    0.0, 315.0,   0.0) },
+{  40, make_orientation("ROLL_90_PITCH_315",           90.0, 315.0,   0.0) },
+{ 100, make_orientation("CUSTOM",                       0.0,   0.0,   0.0) },
 }};
-// [[[end]]] (checksum: 8da0773dfd7a53b866aa7cad6c9ee3d0)
+// [[[end]]] (checksum: 92ae7d7d313456c3645ce7c189d2b6df)
 
 
 std::string to_string(MAV_SENSOR_ORIENTATION orientation)
 {
 	const auto idx = enum_value(orientation);
-	if (idx >= sensor_orientations.size()) {
+	auto it = sensor_orientations.find(idx);
+
+	if (it == sensor_orientations.end()) {
 		ROS_ERROR_NAMED("uas", "SENSOR: wrong orientation index: %d", idx);
 		return std::to_string(idx);
 	}
 
-	return sensor_orientations[idx].first;
+	return it->second.first;
 }
 
 Eigen::Quaterniond sensor_orientation_matching(MAV_SENSOR_ORIENTATION orientation)
 {
-	//const size_t idx(orientation);
-	const auto idx = static_cast<std::underlying_type<MAV_SENSOR_ORIENTATION>::type>(orientation);
-	if (idx >= sensor_orientations.size()) {
+	const auto idx = enum_value(orientation);
+	auto it = sensor_orientations.find(idx);
+
+	if (it == sensor_orientations.end()) {
 		ROS_ERROR_NAMED("uas", "SENSOR: wrong orientation index: %d", idx);
 		return Eigen::Quaterniond::Identity();
 	}
 
-	return sensor_orientations[idx].second;
+	return it->second.second;
 }
 
 int sensor_orientation_from_str(const std::string &sensor_orientation)
@@ -148,16 +152,16 @@ int sensor_orientation_from_str(const std::string &sensor_orientation)
 	// XXX bsearch
 
 	// 1. try to find by name
-	for (size_t idx = 0; idx < sensor_orientations.size(); idx++) {
-		if (sensor_orientations[idx].first == sensor_orientation)
-			return idx;
+	for (const auto& kv : sensor_orientations) {
+		if (kv.second.first == sensor_orientation)
+			return kv.first;
 	}
 
 	// 2. try convert integer
 	// fallback for old configs that uses numeric orientation.
 	try {
 		int idx = std::stoi(sensor_orientation, 0, 0);
-		if (0 > idx || size_t(idx) > sensor_orientations.size()) {
+		if (0 > idx || sensor_orientations.find(idx) == sensor_orientations.end()) {
 			ROS_ERROR_NAMED("uas", "SENSOR: orientation index out of bound: %d", idx);
 			return -1;
 		}

--- a/mavros/src/lib/enum_to_string.cpp
+++ b/mavros/src/lib/enum_to_string.cpp
@@ -153,7 +153,7 @@ std::string to_string(MAV_AUTOPILOT e)
 // to_string_outl(ename)
 // ]]]
 //! MAV_TYPE values
-static const std::array<const std::string, 37> mav_type_strings{{
+static const std::array<const std::string, 36> mav_type_strings{{
 /*  0 */ "Generic micro air vehicle",     // Generic micro air vehicle
 /*  1 */ "Fixed wing aircraft",           // Fixed wing aircraft.
 /*  2 */ "Quadrotor",                     // Quadrotor
@@ -190,7 +190,6 @@ static const std::array<const std::string, 37> mav_type_strings{{
 /* 33 */ "Servo",                         // Servo
 /* 34 */ "Open Drone ID. See https:",     // Open Drone ID. See https://mavlink.io/en/services/opendroneid.html.
 /* 35 */ "Decarotor",                     // Decarotor
-/* 36 */ "Battery",                       // Battery
 }};
 
 std::string to_string(MAV_TYPE e)
@@ -201,14 +200,14 @@ std::string to_string(MAV_TYPE e)
 
 	return mav_type_strings[idx];
 }
-// [[[end]]] (checksum: 8fc2dcd1c935d5d338508685d9eac918)
+// [[[end]]] (checksum: f44b083f4b6be34f26d75692072f09bf)
 
 // [[[cog:
 // ename = 'MAV_TYPE'
 // enum_name_is_value_outl(ename, funcname='to_name', suffix='_names')
 // ]]]
 //! MAV_TYPE values
-static const std::array<const std::string, 37> mav_type_names{{
+static const std::array<const std::string, 36> mav_type_names{{
 /*  0 */ "GENERIC",                       // Generic micro air vehicle
 /*  1 */ "FIXED_WING",                    // Fixed wing aircraft.
 /*  2 */ "QUADROTOR",                     // Quadrotor
@@ -245,7 +244,6 @@ static const std::array<const std::string, 37> mav_type_names{{
 /* 33 */ "SERVO",                         // Servo
 /* 34 */ "ODID",                          // Open Drone ID. See https://mavlink.io/en/services/opendroneid.html.
 /* 35 */ "DECAROTOR",                     // Decarotor
-/* 36 */ "BATTERY",                       // Battery
 }};
 
 std::string to_name(MAV_TYPE e)
@@ -256,7 +254,7 @@ std::string to_name(MAV_TYPE e)
 
 	return mav_type_names[idx];
 }
-// [[[end]]] (checksum: 2d402cd11e54ad35405d53b7ab127e3d)
+// [[[end]]] (checksum: 3775b5a8763b4e66287a471af939ef6c)
 
 // [[[cog:
 // ename = 'MAV_STATE'
@@ -539,7 +537,7 @@ std::string to_string(MAV_FRAME e)
 // suffix = 'MAV_COMP_ID'
 // enum = get_enum(ename)
 //
-// cog.outl(f"static const std::unordered_map<size_t, const std::string> {suffix.lower()}_strings{{{{")
+// cog.outl(f"static const std::unordered_map<typename std::underlying_type<{ename}>::type, const std::string> {suffix.lower()}_strings{{{{")
 // for k, e in enum:
 //     name_short =  e.name[len(suffix) + 1:]
 //     sp = make_whitespace(30, name_short)
@@ -550,7 +548,7 @@ std::string to_string(MAV_FRAME e)
 //
 // cog.outl("}};")
 // ]]]
-static const std::unordered_map<size_t, const std::string> mav_comp_id_strings{{
+static const std::unordered_map<typename std::underlying_type<MAV_COMPONENT>::type, const std::string> mav_comp_id_strings{{
 {   0, "ALL" },                           // Target id (target_component) used to broadcast messages to all components of the receiving system. Components should attempt to process messages with this component ID and forward to components on any other interfaces. Note: This is not a valid *source* component id for a message.
 {   1, "AUTOPILOT1" },                    // System flight controller component ("autopilot"). Only one autopilot is expected in a particular system.
 {  25, "USER1" },                         // Id for a component on privately managed MAVLink network. Can be used for any purpose but may not be published by components outside of the private network.
@@ -660,8 +658,6 @@ static const std::unordered_map<size_t, const std::string> mav_comp_id_strings{{
 { 173, "GIMBAL4" },                       // Gimbal #4
 { 174, "GIMBAL5" },                       // Gimbal #5.
 { 175, "GIMBAL6" },                       // Gimbal #6.
-{ 180, "BATTERY" },                       // Battery #1.
-{ 181, "BATTERY2" },                      // Battery #2.
 { 190, "MISSIONPLANNER" },                // Component that can generate/supply a mission flight plan (e.g. GCS or developer API).
 { 191, "ONBOARD_COMPUTER" },              // Component that lives on the onboard computer (companion computer) and has some generic functionalities, such as settings system parameters and monitoring the status of some processes that don't directly speak mavlink and so on.
 { 195, "PATHPLANNER" },                   // Component that finds an optimal path between points based on a certain constraint (e.g. minimum snap, shortest path, cost, etc.).
@@ -681,11 +677,11 @@ static const std::unordered_map<size_t, const std::string> mav_comp_id_strings{{
 { 242, "TUNNEL_NODE" },                   // Component handling TUNNEL messages (e.g. vendor specific GUI of a component).
 { 250, "SYSTEM_CONTROL" },                // Component for handling system messages (e.g. to ARM, takeoff, etc.).
 }};
-// [[[end]]] (checksum: aa881c50ec1302df3d49fdc6fa6fe13a)
+// [[[end]]] (checksum: 2dc119e3e20f7668a71e8649ba3f67b6)
 
 std::string to_string(MAV_COMPONENT e)
 {
-	size_t idx = enum_value(e);
+	auto idx = enum_value(e);
 	auto it = mav_comp_id_strings.find(idx);
 
 	if (it == mav_comp_id_strings.end())

--- a/mavros/src/mavros/event_launcher.py
+++ b/mavros/src/mavros/event_launcher.py
@@ -189,7 +189,7 @@ class Launcher(object):
             return
 
         # load configuration
-        for k, v in params.iteritems():
+        for k, v in params.items():
             # TODO: add checks for mutually exclusive options
 
             if 'service' in v:

--- a/mavros/src/mavros/param.py
+++ b/mavros/src/mavros/param.py
@@ -193,8 +193,8 @@ def param_get_all(force_pull=False):
     params = rospy.get_param(mavros.get_topic('param'))
 
     return (ret.param_received,
-            sorted((Parameter(k, v) for k, v in params.items()),
-                   key=lambda p: p.param_id)
+            list(sorted((Parameter(k, v) for k, v in params.items()),
+                   key=lambda p: p.param_id))
             )
 
 

--- a/mavros/test/test_sensor_orientation.cpp
+++ b/mavros/test/test_sensor_orientation.cpp
@@ -47,6 +47,14 @@ TEST(UTILS, sensor_orientation_matching__roll_180_yaw_90)
 	EXPECT_QUATERNION(expected, out);
 }
 
+TEST(UTILS, sensor_orientation_matching__custom)
+{
+	auto expected = ftf::quaternion_from_rpy(0.0, 0.0, 0.0);
+	auto out = utils::sensor_orientation_matching(SO::ROTATION_CUSTOM);
+
+	EXPECT_QUATERNION(expected, out);
+}
+
 TEST(UTILS, to_string__none)
 {
 	EXPECT_EQ("NONE", utils::to_string(SO::ROTATION_NONE));
@@ -62,6 +70,11 @@ TEST(UTILS, to_string__roll_180_yaw_90)
 	EXPECT_EQ("ROLL_180_YAW_90", utils::to_string(SO::ROTATION_ROLL_180_YAW_90));
 }
 
+TEST(UTILS, to_string__custom)
+{
+	EXPECT_EQ("CUSTOM", utils::to_string(SO::ROTATION_CUSTOM));
+}
+
 TEST(UTILS, sensor_orientation_from_str__none)
 {
 	EXPECT_EQ(enum_value(SO::ROTATION_NONE), utils::sensor_orientation_from_str("NONE"));
@@ -69,7 +82,7 @@ TEST(UTILS, sensor_orientation_from_str__none)
 
 TEST(UTILS, sensor_orientation_from_str__unknown)
 {
-	EXPECT_LT(utils::sensor_orientation_from_str("completely wrong identificator"), 0);
+	EXPECT_LT(utils::sensor_orientation_from_str("completely wrong identifier"), 0);
 }
 
 TEST(UTILS, sensor_orientation_from_str__number)
@@ -96,6 +109,11 @@ TEST(UTILS, sensor_orientation_from_str__roll_180_yaw_90)
 TEST(UTILS, sensor_orientation_from_str__last_element_roll_90_yaw_270)
 {
 	EXPECT_EQ(enum_value(SO::ROTATION_ROLL_90_YAW_270), utils::sensor_orientation_from_str("ROLL_90_YAW_270"));
+}
+
+TEST(UTILS, sensor_orientation_from_str__custom)
+{
+	EXPECT_EQ(enum_value(SO::ROTATION_CUSTOM), utils::sensor_orientation_from_str("CUSTOM"));
 }
 
 int main(int argc, char **argv)

--- a/mavros_msgs/msg/CommandCode.msg
+++ b/mavros_msgs/msg/CommandCode.msg
@@ -44,11 +44,6 @@ uint16 AIRFRAME_CONFIGURATION = 2520
 uint16 ARM_AUTHORIZATION_REQUEST = 3001            # Request authorization to arm the vehicle to a external entity, the arm authorizer is responsible to request all data that is needs from the vehicle before authorize or deny the request. If approved the progress of command_ack message should be set with period of time that this authorization is valid in seconds or in case it was denied it should be set with one of the reasons in ARM_AUTH_DENIED_REASON.
         
 
-# MAV_CMD_CAMERA
-uint16 CAMERA_TRACK_POINT = 2004                   # If the camera supports point visual tracking (CAMERA_CAP_FLAGS_HAS_TRACKING_POINT is set), this command allows to initiate the tracking.
-uint16 CAMERA_TRACK_RECTANGLE = 2005               # If the camera supports rectangle visual tracking (CAMERA_CAP_FLAGS_HAS_TRACKING_RECTANGLE is set), this command allows to initiate the tracking.
-uint16 CAMERA_STOP_TRACKING = 2010                 # Stops ongoing tracking.
-
 # MAV_CMD_COMPONENT
 uint16 COMPONENT_ARM_DISARM = 400                  # Arms / Disarms a component
 
@@ -58,7 +53,6 @@ uint16 CONDITION_CHANGE_ALT = 113                  # Ascend/descend to target al
 uint16 CONDITION_DISTANCE = 114                    # Delay mission state machine until within desired distance of next NAV point.
 uint16 CONDITION_YAW = 115                         # Reach a certain target angle.
 uint16 CONDITION_LAST = 159                        # NOP - This command is only used to mark the upper limit of the CONDITION commands in the enumeration
-uint16 CONDITION_GATE = 4501                       # Delay mission state machine until gate has been reached.
 
 # MAV_CMD_CONTROL
 uint16 CONTROL_HIGH_LATENCY = 2600                 # Request to start/stop transmitting over the high latency telemetry
@@ -66,7 +60,6 @@ uint16 CONTROL_HIGH_LATENCY = 2600                 # Request to start/stop trans
 # MAV_CMD_DO
 uint16 DO_FOLLOW = 32                              # Begin following a target
 uint16 DO_FOLLOW_REPOSITION = 33                   # Reposition the MAV after a follow target command has been sent
-uint16 DO_ORBIT = 34                               # Start orbiting on the circumference of a circle defined by the parameters. Setting any value NaN results in using defaults.
 uint16 DO_SET_MODE = 176                           # Set system mode.
 uint16 DO_JUMP = 177                               # Jump to the desired command in the mission list.  Repeat this action only the specified number of times
 uint16 DO_CHANGE_SPEED = 178                       # Change speed and/or throttle set points.
@@ -78,17 +71,16 @@ uint16 DO_SET_SERVO = 183                          # Set a servo to a desired PW
 uint16 DO_REPEAT_SERVO = 184                       # Cycle a between its nominal setting and a desired PWM for a desired number of cycles with a desired period.
 uint16 DO_FLIGHTTERMINATION = 185                  # Terminate flight immediately
 uint16 DO_CHANGE_ALTITUDE = 186                    # Change altitude set point.
-uint16 DO_SET_ACTUATOR = 187                       # Sets actuators (e.g. servos) to a desired value. The actuator numbers are mapped to specific outputs (e.g. on any MAIN or AUX PWM or UAVCAN) using a flight-stack specific mechanism (i.e. a parameter).
 uint16 DO_LAND_START = 189                         # Mission command to perform a landing. This is used as a marker in a mission to tell the autopilot where a sequence of mission items that represents a landing starts. It may also be sent via a COMMAND_LONG to trigger a landing, in which case the nearest (geographically) landing sequence in the mission will be used. The Latitude/Longitude is optional, and may be set to 0 if not needed. If specified then it will be used to help find the closest landing sequence.
 uint16 DO_RALLY_LAND = 190                         # Mission command to perform a landing from a rally point.
 uint16 DO_GO_AROUND = 191                          # Mission command to safely abort an autonomous landing.
 uint16 DO_REPOSITION = 192                         # Reposition the vehicle to a specific WGS84 global position.
 uint16 DO_PAUSE_CONTINUE = 193                     # If in a GPS controlled position mode, hold the current position or continue.
 uint16 DO_SET_REVERSE = 194                        # Set moving direction to forward or reverse.
-uint16 DO_SET_ROI_LOCATION = 195                   # Sets the region of interest (ROI) to a location. This can then be used by the vehicle's control system to control the vehicle attitude and the attitude of various sensors such as cameras. This command can be sent to a gimbal manager but not to a gimbal device. A gimbal is not to react to this message.
-uint16 DO_SET_ROI_WPNEXT_OFFSET = 196              # Sets the region of interest (ROI) to be toward next waypoint, with optional pitch/roll/yaw offset. This can then be used by the vehicle's control system to control the vehicle attitude and the attitude of various sensors such as cameras. This command can be sent to a gimbal manager but not to a gimbal device. A gimbal device is not to react to this message.
-uint16 DO_SET_ROI_NONE = 197                       # Cancels any previous ROI command returning the vehicle/sensors to default flight characteristics. This can then be used by the vehicle's control system to control the vehicle attitude and the attitude of various sensors such as cameras. This command can be sent to a gimbal manager but not to a gimbal device. A gimbal device is not to react to this message. After this command the gimbal manager should go back to manual input if available, and otherwise assume a neutral position.
-uint16 DO_SET_ROI_SYSID = 198                      # Mount tracks system with specified system ID. Determination of target vehicle position may be done with GLOBAL_POSITION_INT or any other means. This command can be sent to a gimbal manager but not to a gimbal device. A gimbal device is not to react to this message.
+uint16 DO_SET_ROI_LOCATION = 195                   # Sets the region of interest (ROI) to a location. This can then be used by the vehicle's control system to control the vehicle attitude and the attitude of various sensors such as cameras.
+uint16 DO_SET_ROI_WPNEXT_OFFSET = 196              # Sets the region of interest (ROI) to be toward next waypoint, with optional pitch/roll/yaw offset. This can then be used by the vehicle's control system to control the vehicle attitude and the attitude of various sensors such as cameras.
+uint16 DO_SET_ROI_NONE = 197                       # Cancels any previous ROI command returning the vehicle/sensors to default flight characteristics. This can then be used by the vehicle's control system to control the vehicle attitude and the attitude of various sensors such as cameras.
+uint16 DO_SET_ROI_SYSID = 198                      # Mount tracks system with specified system ID. Determination of target vehicle position may be done with GLOBAL_POSITION_INT or any other means.
 uint16 DO_CONTROL_VIDEO = 200                      # Control onboard camera system.
 uint16 DO_SET_ROI = 201                            # Sets the region of interest (ROI) for a sensor set or the vehicle itself. This can then be used by the vehicle's control system to control the vehicle attitude and the attitude of various sensors such as cameras.
 uint16 DO_DIGICAM_CONFIGURE = 202                  # Configure digital camera. This is a fallback message for systems that have not yet implemented PARAM_EXT_XXX messages and camera definition files (see https://mavlink.io/en/services/camera_def.html ).
@@ -109,12 +101,10 @@ uint16 DO_GUIDED_LIMITS = 222                      # Set limits for external con
 uint16 DO_ENGINE_CONTROL = 223                     # Control vehicle engine. This is interpreted by the vehicles engine controller to change the target engine state. It is intended for vehicles with internal combustion engines
 uint16 DO_SET_MISSION_CURRENT = 224                # Set the mission item with sequence number seq as current item. This means that the MAV will continue to this mission item on the shortest path (not following the mission items in-between).
 uint16 DO_LAST = 240                               # NOP - This command is only used to mark the upper limit of the DO commands in the enumeration
-uint16 DO_UPGRADE = 247                            # Request a target system to start an upgrade of one (or all) of its components. For example, the command might be sent to a companion computer to cause it to upgrade a connected flight controller. The system doing the upgrade will report progress using the normal command protocol sequence for a long running operation. Command protocol information: https://mavlink.io/en/services/command.html.
 uint16 DO_JUMP_TAG = 601                           # Jump to the matching tag in the mission list. Repeat this action for the specified number of times. A mission should contain a single matching tag for each jump. If this is not the case then a jump to a missing tag should complete the mission, and a jump where there are multiple matching tags should always select the one with the lowest mission sequence number.
-uint16 DO_GIMBAL_MANAGER_PITCHYAW = 1000           # High level setpoint to be sent to a gimbal manager to set a gimbal attitude. It is possible to set combinations of the values below. E.g. an angle as well as a desired angular rate can be used to get to this angle at a certain angular rate, or an angular rate only will result in continuous turning. NaN is to be used to signal unset. Note: a gimbal is never to react to this command but only the gimbal manager.
-uint16 DO_GIMBAL_MANAGER_CONFIGURE = 1001          # Gimbal configuration to set which sysid/compid is in primary and secondary control.
 uint16 DO_TRIGGER_CONTROL = 2003                   # Enable or disable on-board camera triggering system.
 uint16 DO_VTOL_TRANSITION = 3000                   # Request VTOL transition
+uint16 DO_ADSB_OUT_IDENT = 10001                   # Trigger the start of an ADSB-out IDENT. This should only be used when requested to do so by an Air Traffic Controller in controlled airspace. This starts the IDENT which is then typically held for 18 seconds by the hardware per the Mode A, C, and S transponder spec.
 uint16 DO_WINCH = 42600                            # Command to operate winch.
 
 # MAV_CMD_FIXED
@@ -124,15 +114,9 @@ uint16 FIXED_MAG_CAL_YAW = 42006                   # Magnetometer calibration ba
 uint16 GET_HOME_POSITION = 410                     # Request the home position from the vehicle.
 uint16 GET_MESSAGE_INTERVAL = 510                  # Request the interval between messages for a particular MAVLink message ID. The receiver should ACK the command and then emit its response in a MESSAGE_INTERVAL message.
 
-# MAV_CMD_ILLUMINATOR
-uint16 ILLUMINATOR_ON_OFF = 405                    # Turns illuminators ON/OFF. An illuminator is a light source that is used for lighting up dark areas external to the sytstem: e.g. a torch or searchlight (as opposed to a light source for illuminating the system itself, e.g. an indicator light).
-
 # MAV_CMD_IMAGE
 uint16 IMAGE_START_CAPTURE = 2000                  # Start image capture sequence. Sends CAMERA_IMAGE_CAPTURED after each capture. Use NaN for reserved values.
 uint16 IMAGE_STOP_CAPTURE = 2001                   # Stop image capture sequence Use NaN for reserved values.
-
-# MAV_CMD_INJECT
-uint16 INJECT_FAILURE = 420                        # Inject artificial failure for testing purposes. Note that autopilots should implement an additional protection before accepting this command such as a specific param setting.
 
 # MAV_CMD_JUMP
 uint16 JUMP_TAG = 600                              # Tagged jump target. Can be jumped to with MAV_CMD_DO_JUMP_TAG.
@@ -148,7 +132,7 @@ uint16 MISSION_START = 300                         # start running a mission
 uint16 NAV_WAYPOINT = 16                           # Navigate to waypoint.
 uint16 NAV_LOITER_UNLIM = 17                       # Loiter around this waypoint an unlimited amount of time
 uint16 NAV_LOITER_TURNS = 18                       # Loiter around this waypoint for X turns
-uint16 NAV_LOITER_TIME = 19                        # Loiter at the specified latitude, longitude and altitude for a certain amount of time. Multicopter vehicles stop at the point (within a vehicle-specific acceptance radius). Forward-only moving vehicles (e.g. fixed-wing) circle the point with the specified radius/direction. If the Heading Required parameter (2) is non-zero forward moving aircraft will only leave the loiter circle once heading towards the next waypoint.
+uint16 NAV_LOITER_TIME = 19                        # Loiter around this waypoint for X seconds
 uint16 NAV_RETURN_TO_LAUNCH = 20                   # Return to launch location
 uint16 NAV_LAND = 21                               # Land at location.
 uint16 NAV_TAKEOFF = 22                            # Takeoff from ground / hand. Vehicles that support multiple takeoff modes (e.g. VTOL quadplane) should take off using the currently configured mode.
@@ -160,7 +144,7 @@ uint16 NAV_LOITER_TO_ALT = 31                      # Begin loiter at the specifi
 uint16 NAV_ROI = 80                                # Sets the region of interest (ROI) for a sensor set or the vehicle itself. This can then be used by the vehicle's control system to control the vehicle attitude and the attitude of various sensors such as cameras.
 uint16 NAV_PATHPLANNING = 81                       # Control autonomous path planning on the MAV.
 uint16 NAV_SPLINE_WAYPOINT = 82                    # Navigate to waypoint using a spline path.
-uint16 NAV_VTOL_TAKEOFF = 84                       # Takeoff from ground using VTOL mode, and transition to forward flight with specified heading. The command should be ignored by vehicles that dont support both VTOL and fixed-wing flight (multicopters, boats,etc.).
+uint16 NAV_VTOL_TAKEOFF = 84                       # Takeoff from ground using VTOL mode, and transition to forward flight with specified heading.
 uint16 NAV_VTOL_LAND = 85                          # Land using VTOL mode
 uint16 NAV_GUIDED_ENABLE = 92                      # hand control over to an external controller
 uint16 NAV_DELAY = 93                              # Delay the next navigation command a number of seconds or until a specified time
@@ -188,9 +172,6 @@ uint16 OVERRIDE_GOTO = 252                         # Override current mission wi
 # MAV_CMD_PANORAMA
 uint16 PANORAMA_CREATE = 2800                      # Create a panorama at the current position
 
-# MAV_CMD_PARAM
-uint16 PARAM_TRANSACTION = 900                     # Request to start or end a parameter transaction. Multiple kinds of transport layers can be used to exchange parameters in the transaction (param, param_ext and mavftp). The command response can either be a success/failure or an in progress in case the receiving side takes some time to apply the parameters.
-
 # MAV_CMD_PAYLOAD
 uint16 PAYLOAD_PREPARE_DEPLOY = 30001              # Deploy payload on a Lat / Lon / Alt position. This includes the navigation to reach the required release position and velocity.
 uint16 PAYLOAD_CONTROL_DEPLOY = 30002              # Control the payload deployment.
@@ -211,18 +192,18 @@ uint16 REQUEST_CAMERA_SETTINGS = 522               # Request camera settings (CA
 uint16 REQUEST_STORAGE_INFORMATION = 525           # Request storage information (STORAGE_INFORMATION). Use the command's target_component to target a specific component's storage.
 uint16 REQUEST_CAMERA_CAPTURE_STATUS = 527         # Request camera capture status (CAMERA_CAPTURE_STATUS)
 uint16 REQUEST_FLIGHT_INFORMATION = 528            # Request flight information (FLIGHT_INFORMATION)
-uint16 REQUEST_CAMERA_IMAGE_CAPTURE = 2002         # Re-request a CAMERA_IMAGE_CAPTURED message.
 uint16 REQUEST_VIDEO_STREAM_INFORMATION = 2504     # Request video stream information (VIDEO_STREAM_INFORMATION)
 uint16 REQUEST_VIDEO_STREAM_STATUS = 2505          # Request video stream status (VIDEO_STREAM_STATUS)
 
 # MAV_CMD_RESET
 uint16 RESET_CAMERA_SETTINGS = 529                 # Reset all camera settings to Factory Default
 
+# MAV_CMD_RUN
+uint16 RUN_PREARM_CHECKS = 401                     # Instructs system to run pre-arm checks.  This command should return MAV_RESULT_TEMPORARILY_REJECTED in the case the system is armed, otherwse MAV_RESULT_ACCEPTED.  Note that the return value from executing this command does not indicate whether the vehicle is armable or not, just whether the system has successfully run/is currently running the checks.  The result of the checks is reflected in the SYS_STATUS message.
+
 # MAV_CMD_SET
 uint16 SET_MESSAGE_INTERVAL = 511                  # Set the interval between messages for a particular MAVLink message ID. This interface replaces REQUEST_DATA_STREAM.
 uint16 SET_CAMERA_MODE = 530                       # Set camera running mode. Use NaN for reserved values. GCS will send a MAV_CMD_REQUEST_VIDEO_STREAM_STATUS command after a mode change if the camera supports video streaming.
-uint16 SET_CAMERA_ZOOM = 531                       # Set camera zoom. Camera must respond with a CAMERA_SETTINGS message (on success).
-uint16 SET_CAMERA_FOCUS = 532                      # Set camera focus. Camera must respond with a CAMERA_SETTINGS message (on success).
 uint16 SET_GUIDED_SUBMODE_STANDARD = 4000          # This command sets the submode to standard guided when vehicle is in guided mode. The vehicle holds position and altitude and the user can input the desired velocities along all three axes.
                   
 uint16 SET_GUIDED_SUBMODE_CIRCLE = 4001            # This command sets submode circle when vehicle is in guided mode. Vehicle flies along a circle facing the center of the circle. The user can input the velocity along the circle and change the radius. If no input is given the vehicle will hold position.
@@ -243,4 +224,4 @@ uint16 VIDEO_STOP_CAPTURE = 2501                   # Stop the current video capt
 uint16 VIDEO_START_STREAMING = 2502                # Start video streaming
 uint16 VIDEO_STOP_STREAMING = 2503                 # Stop the given video stream
 
-# [[[end]]] (checksum: d1339844f21a28ab73ddcc1ad911802f)
+# [[[end]]] (checksum: 01fa1723eb3702c301c5562fdf293bbc)

